core-contracts/contracts/bns.clar

```diff
@@ -0,0 +1,958 @@
+
+;; NAME_REGISTRATION
+;; This is the second transaction to be sent. It reveals the salt and the name to all BNS nodes,
+;; and assigns the name an initial public key hash and zone file hash
+(define-public (name-register (namespace (buff 19))
+                              (name (buff 16))
+                              (salt (buff 20))
+                              (zonefile-content (buff 40960)))
+  (let (
+    (hashed-salted-fqn (hash160 (concat (concat (concat name ".") namespace) salt)))
+    (name-currently-owned (map-get owner-name ((owner contract-caller)))))
+    (let ( 
+        (preorder (expects!
+          (map-get name-preorders ((hashed-salted-fqn hashed-salted-fqn) (buyer contract-caller)))
+          (err err-name-preorder-not-found)))
+        (namespace-props (expects!
+          (map-get namespaces ((namespace namespace)))
+          (err err-namespace-not-found)))
+        (current-owner (nft-get-owner names (tuple (name name) (namespace namespace))))
+        (can-sender-register-name (if (is-none? name-currently-owned)
+                                 'true
+                                  (expects! 
+                                    (is-name-lease-expired?
+                                      (expects! (get namespace name-currently-owned) (err err-panic))
+                                      (expects! (get name name-currently-owned) (err err-panic)))
+                                    (err err-panic)))))
+      ;; The name must only have valid chars
+      (asserts!
+        (not (has-invalid-chars name))
+        (err err-name-charset-invalid))
+      ;; The name must not exist yet, or be expired
+      (if (is-none? current-owner)
+        'true
+        (asserts!
+          (expects! (is-name-lease-expired? namespace name) (err err-panic))
+          (err err-name-unavailable)))
+      ;; The name's namespace must be launched
+      (asserts!
+        (eq? (is-none? (get launched-at namespace-props)) 'false)
+        (err err-namespace-not-launched))
+      ;; The preorder must have been created after the launch of the namespace
+      (asserts!
+        (> (get created-at preorder) (expects! (get launched-at namespace-props) (err err-panic)))
+        (err err-name-preordered-before-namespace-launch))
+      ;; The preorder entry must be unclaimed - todo(ludo): is this assertion redundant?
+      (asserts!
+        (eq? (get claimed preorder) 'false)
+        (err err-name-already-claimed))
+      ;; Less than 24 hours must have passed since the name was preordered
+      (asserts!
+        (< block-height (+ (get created-at preorder) name-preorder-claimability-ttl))
+        (err err-name-claimability-expired))
+      ;; The amount burnt must be equal to or greater than the cost of the name
+      (asserts!
+        (>= (get stx-burned preorder) (compute-name-price name (get price-function namespace-props)))
+        (err err-name-stx-burnt-insufficient))
+      ;; The principal can register a name
```

Comment:
Why enforce one name per principal? The current BNS implementation sets this limit at 20 names per address, I think.

Reply:
This was a feature request coming from Jude.

Reply:
@jcnelson -- this seems like a good idea in isolation, but won't this be a problem for forking from the current name ownership set?

Reply:
On further thought, I'm not sure if this is a good idea, because if someone wants a _contract principal_ to be able to own names, it seems like the contract should be able to own multiple names.

Reply:
This is a good point -- if we want contracts to own names, then this restriction should be lifted for contract principals at least.

My desire to restrict it to at most one name per standard principal is to deal with the fact that blockstack.js, the browser, and apps are fundamentally unable to determine which name should be used for a given ID-address.  For example, if an ID-address has two names, the tooling must prompt the user as to which name they'd like to expose to the app.  We have the chance to address this by removing the "one address can own 25 names" constraint in the current BNS system.

We can make it so that addresses owning 2+ names will continue to do so when the name db is imported to this BNS smart contract, but I think we should stop allowing standard principals to register or renew multiple names going forward.

Reply:
> contract should be able to own multiple names

We'd need a new native func for testing if a principal is standard or contract.